### DISCUSSION TITLE
Logging from within custom rule

### DISCUSSION THREAD

**Discussion by l3ender** [eslint/eslint#18707] [Q&A] [answered]
We have custom rules which we've added to our project, and I'm wondering if there's a preferred way to log output during rule execution. Until now we've been using `console.log` in our rules, but it can't be turned on/off (e.g. with `--debug` flag), and the output isn't styled/formatted the way other debug logs are shown (in color, with name of module which logged).

I did some looking and didn't see any sort of logging framework which can be utilized within a rule. Apologies if I missed it!

Styled/formatted output when using `--debug`:

![image](https://github.com/user-attachments/assets/5d0e8087-3f24-4ac3-a58b-08d28ada82e8)

**Comment by amareshsm** [comment]
One way of controlling logging and disabling it when not needed, you use a condition or a configuration flag. Here’s a method to control logging with a configuration flag:

1. Define a flag to enable or disable logging.
2. Use this flag to conditionally execute console.log statements.

```js
module.exports = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Log data for debugging',
      category: 'Best Practices',
      recommended: false,
    },
    schema: [], // No options
  },
  create(context) {
    const enableLogging = process.env.ESLINT_LOGGING === 'true';

    return {
      Identifier(node) {
        if (enableLogging) {
          console.log('Identifier node:', node);
        }
      },
    };
  },
};

```

Set Up Environment Variable:

```bash
export ESLINT_LOGGING=true
```

You can use any logging libraries with the above approach to style and format the output.

**Reply by l3ender** [reply]
Thanks for the response, I appreciate it!

Apart from rolling our own logging mechanism to turn on/off, is there a way hook into the library's logging. Aside from needing to handle enable/disable logic, this approach doesn't format output with styling/indentation which matches the library or provide [what we believe are] timings for individual operations (far right ms value for each command):

![image](https://github.com/user-attachments/assets/3920add3-2503-4f2a-8ad3-7b5a7b96493c)

**Reply by nzakas** [reply, marked as answer]
No, there is no way to do this.
